<h2>Introduction</h2>
<p>A paired samples t-test is used to compare the means of two samples when each observation in one sample can be paired with an observation in the other sample.</p>

<p>This tutorial explains how to conduct a paired samples t-test in Python.</p>

<h2>Example: Paired Samples T-Test in Python</h2>
<p>Suppose we want to know whether a certain study program significantly impacts student performance on a particular exam. To test this, we have 15 students in a class take a pre-test. Then, we have each of the students participate in the study program for two weeks. Then, the students retake a test of similar difficulty.</p>

<p>To compare the difference between the mean scores on the first and second test, we use a paired samples t-test because for each student their first test score can be paired with their second test score.</p>

<p>Perform the following steps to conduct a paired samples t-test in Python.</p>

<h3>Step 1: Create the data.</h3>
<p>First, we’ll create two arrays to hold the pre and post-test scores:</p>

In [1]:
pre = [88, 82, 84, 93, 75, 78, 84, 87, 95, 91, 83, 89, 77, 68, 91]
post = [91, 84, 88, 90, 79, 80, 88, 90, 90, 96, 88, 89, 81, 74, 92]

<h3>Step 2: Conduct a Paired Samples T-Test.</h3>

<p>Next, we’ll use the ttest_rel() function from the scipy.stats library to conduct a paired samples t-test, which uses the following syntax:</p>
<pre>
ttest_rel(a, b)
</pre>
<p>where:</p>
<ul>
  <li>a: an array of sample observations from group 1</li>
  <li>b: an array of sample observations from group 2</li>
</ul>
<p>Here’s how to use this function in our specific example:</p>

In [2]:
import scipy.stats as stats
#perform the paired samples t-test
stats.ttest_rel(pre, post)

Ttest_relResult(statistic=-2.9732484231168796, pvalue=0.01007144862643272)

<p>The test statistic is -2.9732 and the corresponding two-sided p-value is 0.0101.</p>

<h3>Step 3: Interpret the results.</h3>

<p>In this example, the paired samples t-test uses the following null and alternative hypotheses:</p>
<ul>
   <li>H0: The mean pre-test and post-test scores are equal</li>

   <li>HA:The mean pre-test and post-test scores are not equal</li>
</ul>
<p>Since the p-value (0.0101) is less than 0.05, we reject the null hypothesis. We have sufficient evidence to say that the true mean test score is different for students before and after participating in the study program.</p>

<h2>How to Perform a Shapiro-Wilk Test</h2>
<p>The Shapiro-Wilk test is a test of normality. It is used to determine whether or not a sample comes from a normal distribution.</p>

<p>To perform a Shapiro-Wilk test in Python we can use the scipy.stats.shapiro() function, which takes on the following syntax:</p>
<pre>
scipy.stats.shapiro(x)
</pre>
<p>where:</p>
<p>x: An array of sample data.</p>
<p>This function returns a test statistic and a corresponding p-value. If the p-value is below a certain significance level, then we have sufficient evidence to say that the sample data does not come from a normal distribution.</p>

<p>This tutorial shows a couple examples of how to use this function in practice.</p>
<p>The following code shows how to perform a Shapiro-Wilk test on pre data to determine if it came from a normal distribution:</p>

In [3]:
import scipy.stats as stats
#perform Shapiro-Wilk test
stats.shapiro(pre)

ShapiroResult(statistic=0.9618234634399414, pvalue=0.7240561842918396)

<p>From the output we can see that the test statistic is 0.9618 and the corresponding p-value is 0.7241.</p>

<p>Since the p-value is not less than 0.05, we fail to reject the null hypothesis. We do not have sufficient evidence to say that the sample data does not come from a normal distribution.</p>

In [4]:
stats.shapiro(post)

ShapiroResult(statistic=0.9243801832199097, pvalue=0.224565327167511)

<p>From the output we can see that the test statistic is 0.924 and the corresponding p-value is 0.2246.</p>

<p>Since the p-value is not less than 0.05, we fail to reject the null hypothesis. We do not have sufficient evidence to say that the sample data does not come from a normal distribution.</p>

<h2>How to Perform Bartlett’s Test</h2>
<p>Bartlett’s test is a statistical test that is used to determine whether or not the variances between several groups are equal.</p>

<p>Many statistical tests (like a one-way ANOVA) assume that variances are equal across samples. Bartlett’s test can be used to verify that assumption. This test uses the following null and alternative hypotheses:</p>
<ul>
   <li>H0: The variance among each group is equal.</li>
   <li>HA: At least one group has a variance that is not equal to the rest.</li>
</ul>
<p>The test statistic follows a Chi-Square distribution with k-1 degrees of freedom where k is the number of groups.</p>

<p>If the corresponding p-value of the test statistic is less than some significance level (like α = .05) then we can reject the null hypothesis and conclude that not all groups have the same variance.</p>

<p>The following step-by-step example explains how to perform Bartlett’s test in Python.</p>

In [5]:
import scipy.stats as stats

#perform Bartlett's test 
stats.bartlett(pre,post)

BartlettResult(statistic=0.7608658798370089, pvalue=0.3830576835574586)

<p>The test returns the following results:</p>
<ul>
   <li>Test statistic B: 0.7609</li>
   <li>P-value: 0.3831</li>
</ul>

<p>Since the p-value is not less than 0.05, we will fail to reject the null hypothesis. In other words, we don’t have sufficient evidence to say that the two groups have different variances.</p>